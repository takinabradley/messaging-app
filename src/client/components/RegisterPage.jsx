import {
  redirect,
  Form,
  useParams,
  useSearchParams,
  useActionData
} from "react-router-dom"

import auth from "../modules/auth"

export async function loader() {
  if (await auth.isLoggedIn) return redirect("/app")
  return null
}

export async function action({ request }) {
  const data = await request.formData()

  const registerData = await auth.register(
    data.get("username"),
    data.get("password"),
    data.get("verifyPassword")
  )

  if (registerData.success) {
    return redirect("/app")
  } else {
    return { errors: registerData.errors }
  }
}

export default function RegisterPage() {
  const actionData = useActionData()

  return (
    <div className="RegisterPage" data-testid="RegisterPage">
      Please register...
      <Form action="/register" method="POST" name="RegisterPage__form" id="RegisterPage__form">
        <label htmlFor="RegisterPage__username">
          username
          <input type="text" name="username" id="RegisterPage__username"/>
        </label>

        <label htmlFor="RegisterPage__password">
          password
          <input type="password" name="password" id="RegisterPage__password"/>
        </label>
        <label htmlFor="RegisterPage__verifyPassword" >
          Confirm Password:
          <input type="password" name="verifyPassword" id="RegisterPage__verifyPassword"/>
        </label>
        <button>register</button>
      </Form>
      <output form="RegisterPage__form">
        {actionData && actionData.errors
          ? actionData.errors.map((err) => (
              <div key={err.msg}>
                {err.path}: {err.msg}
              </div>
            ))
          : null}
      </output>
    </div>
  )
}
