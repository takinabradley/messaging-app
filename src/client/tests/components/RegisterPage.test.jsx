import { describe, expect, it, test } from "vitest";
import RegisterPage, {loader as RegisterPageLoader, action as RegisterPageAction} from "../../components/RegisterPage.jsx";
import { render, screen } from "@testing-library/react";
import { createMemoryRouter, RouterProvider } from "react-router-dom"
import userEvent from "@testing-library/user-event";

describe("RegisterPage", () => {
  test("renders", async () => {
    const router = createMemoryRouter([
      {
        path: "/register",
        element: <RegisterPage />,
      }
    ], {initialEntries: ["/", "/register"], initialIndex: 1})

    render(<RouterProvider router={router} />)

    expect(screen.getByTestId("RegisterPage")).toBeInTheDocument()    
    expect(screen.getByTestId("RegisterPage")).toMatchSnapshot()

  })

  test("output element shows an error from form action on submit", async () => {
    const user = userEvent.setup()
    const router = createMemoryRouter([
      {
        path: "/register",
        element: <RegisterPage />,
        action: () => ({errors: [{path: "unknown", msg: "something went wrong!"}]})
      }
    ], {initialEntries: ["/", "/register"], initialIndex: 1})

    render(<RouterProvider router={router} />)

    const registerButton = screen.getByRole("button", {name: "register"})
    await user.click(registerButton)

    expect(screen.getByRole("status").childElementCount).toBe(1)
    expect(screen.getByRole("status").children[0]).toHaveTextContent("unknown: something went wrong!")
  })

  test("output element shows multiple errors from the form action", async () => {
    const user = userEvent.setup()
    const router = createMemoryRouter([
      {
        path: "/register",
        element: <RegisterPage />,
        action: () => ({
            errors: [
                {path: "unknown", msg: "something went wrong!"}, 
                {path: "username", msg: "username already taken!"}
            ]
        })
      }
    ], {initialEntries: ["/", "/register"], initialIndex: 1})

    render(<RouterProvider router={router} />)


    const registerButton = screen.getByRole("button", {name: "register"})
    await user.click(registerButton)

    expect(screen.getByRole("status").childElementCount).toBe(2)
    expect(screen.getByRole("status").children[0]).toHaveTextContent("unknown: something went wrong!")
    expect(screen.getByRole("status").children[1]).toHaveTextContent("username: username already taken!")
  })

  test("output element shows no error from form action on submit when there is an empty list of errors", async () => {
    const user = userEvent.setup()
    const router = createMemoryRouter([
      {
        path: "/register",
        element: <RegisterPage />,
        action: () => ({errors: []})
      }
    ], {initialEntries: ["/", "/register"], initialIndex: 1})

    render(<RouterProvider router={router} />)

    const registerButton = screen.getByRole("button", {name: "register"})
    await user.click(registerButton)

    expect(screen.getByRole("status").childElementCount).toBe(0)
  })

  test("doesn't break when action returns nothing", async () => {
    const user = userEvent.setup()
    const router = createMemoryRouter([
      {
        path: "/register",
        element: <RegisterPage />,
        action: () => null
      }
    ], {initialEntries: ["/", "/register"], initialIndex: 1})

    render(<RouterProvider router={router} />)

    const registerButton = screen.getByRole("button", {name: "register"})
    await user.click(registerButton)

    expect(screen.getByTestId("RegisterPage")).toBeInTheDocument()
    expect(screen.getByRole("status").childElementCount).toBe(0)
  })

  test("doesn't break when action returns empty errors field", async () => {
    const user = userEvent.setup()
    const router = createMemoryRouter([
      {
        path: "/register",
        element: <RegisterPage />,
        action: () => ({errors: undefined})
      }
    ], {initialEntries: ["/", "/register"], initialIndex: 1})

    render(<RouterProvider router={router} />)

    const registerButton = screen.getByRole("button", {name: "register"})
    await user.click(registerButton)

    expect(screen.getByTestId("RegisterPage")).toBeInTheDocument()
    expect(screen.getByRole("status").childElementCount).toBe(0)
  })

  test("users are able to submit username, password, and verifyPassword fields to form action", async () => {
    const user = userEvent.setup()

    let resolveUserCredentials = null
    const userCredentials = new Promise((res, rej) => {
      resolveUserCredentials = res
    })

    const router = createMemoryRouter([
      {
        path: "/register",
        element: <RegisterPage />,
        action: async ({request}) => {
          const data = await request.formData()
          resolveUserCredentials({username: data.get("username"), password: data.get("password"), verifyPassword: data.get("verifyPassword")})
          return null
        }
      }
    ], {initialEntries: ["/", "/register"], initialIndex: 1})

    render(<RouterProvider router={router} />)

    const usernameField = screen.getByRole("textbox", {name: "username"})
    const passwordField = screen.getByLabelText("password")
    const verifyPasswordField = screen.getByLabelText("Confirm Password:")
    const registerButton = screen.getByRole("button", {name: "register"})

    await user.type(usernameField, "hello")
    await user.type(passwordField, "there")
    await user.type(verifyPasswordField, "there")
    await user.click(registerButton)

    expect((await userCredentials).username).toBe("hello")
    expect((await userCredentials).password).toBe("there")
    expect((await userCredentials).verifyPassword).toBe("there")
  })
})