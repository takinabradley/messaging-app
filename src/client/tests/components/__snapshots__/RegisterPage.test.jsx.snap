// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RegisterPage > renders 1`] = `
<div
  class="RegisterPage"
  data-testid="RegisterPage"
>
  Please register...
  <form
    action="/register"
    id="RegisterPage__form"
    method="post"
    name="RegisterPage__form"
  >
    <label
      for="RegisterPage__username"
    >
      username
      <input
        id="RegisterPage__username"
        name="username"
        type="text"
      />
    </label>
    <label
      for="RegisterPage__password"
    >
      password
      <input
        id="RegisterPage__password"
        name="password"
        type="password"
      />
    </label>
    <label
      for="RegisterPage__verifyPassword"
    >
      Confirm Password:
      <input
        id="RegisterPage__verifyPassword"
        name="verifyPassword"
        type="password"
      />
    </label>
    <button>
      register
    </button>
  </form>
  <output
    form="RegisterPage__form"
  />
</div>
`;
